CNDB-12308: Remove index-based sorting at CQL level

adelapena opened the pull request:
It was recently found that index-based sorting is done twice in the coordinator. The first sorting is done by `StorageAttachedIndexQueryPlan#postProcessor`, at the service level, which sorts a `PartitionIterator`. Then a second sorting is done by `SortedRowsBuilder`, at the CQL level, which sorts a simpler `List<ByteBuffer>`.

[CNDB-11762](https://github.com/riptano/cndb/issues/11762) removed the entire `StorageAttachedIndexQueryPlan#postProcessor` method so sorting was not duplicated anymore. However, this was problematic for implementations of `QueryInfoTracker.ReadTracker`, which is applied after sorting with the postprocessor, but before sorting at the CQL level, [here](https://github.com/datastax/cassandra/blob/main/src/java/org/apache/cassandra/service/StorageProxy.java#L2304-L2310). Thus, that change was reverted to avoid breaking external trackers relying on the previous behaviour.

I think what we should do instead is to get rid of index-based sorting at the CQL level, so all sorting is done by the postprocessor. The postprocessor works at a lower level so, with some adjustments, it should be able to do all the sorting on its own without needing CQL intervention. That should remove the duplicated sorting and simplify CQL a bit.

### Checklist before you submit for review
- [x] Make sure there is a PR in the CNDB project updating the Converged Cassandra version
- [x] Use `NoSpamLogger` for log lines that may appear frequently in the logs
- [x] Verify test results on Butler
- [x] Test coverage for new/modified code is > 80%
- [x] Proper code formatting
- [x] Proper title for each commit staring with the project-issue number, like CNDB-1234
- [x] Each commit has a meaningful description
- [x] Each commit is not very long and contains related changes
- [x] Renames, moves and reformatting are in distinct commits

adelapena commented:
@michaeljmarshall could you please review this one?

pkolaczk commented on src/java/org/apache/cassandra/index/sai/plan/TopKProcessor.java:
```diff
@@ -244,7 +269,7 @@ private <U extends Unfiltered, R extends BaseRowIterator<U>, P extends BaseParti
                         topK.addAll(pr.rows);
-                        for (var uf: pr.tombstones)
+                        for (var uf : pr.tombstones)
                             addUnfiltered(unfilteredByPartition, pr.partitionInfo, uf);
```
It's not obvious to me that `unfilteredByPartition != null` here. There is no null check above and it looks like this whole `else` branch may run if `partitions instanceof PartitionIterator` in which case we don't initialize `unfilteredByPartition`. Is it a bug or am I missing something? If not a bug, can you explain why NPE won't happen here?

adelapena replied:
`unfilteredByPartition` can be null only if we are sorting a `PartitionIterator`. I think in that case the `PartitionResults` shouldn't contain any tombstones, so there won't be any iteration in this loop. I can anyway add an `if` or an assertion if it makes it clearer/safer.

pkolaczk replied:
Yeah, please add an assertion.

adelapena replied:
+1, just added an assertion.

adelapena commented:
I have rebased the branch and replaced the TODO comment in `TopKProcessor` by a comment trying to reproduce [this explanation](https://github.com/riptano/cndb/issues/12308#issuecomment-2581273037) by @michaeljmarshall .

pkolaczk reviewed (approved):
Looks good and makes sense. 
It is a bit hard to understand the logic of `filterInternal` which seems to be called both in the context of the coordinator and the replica, and has different sorting depending on that context. I wonder if it wouldn't be cleaner to separate them into two methods? But I don't have a strong opinion on this.

adelapena commented:
@pkolaczk thanks for the review. I have pushed a last commit dividing `TopKProcessor#filter` into two separate methods, as suggested. Please let me know if it looks good. Maybe `TopKProcessor#processPartitions` could use some further refactoring, but I'd prefer to leave that for a separate ticket.

michaeljmarshall reviewed (approved):
LGTM. I specifically paid close attention to the `TopKProcessor`. I'll clean it up more soon when I get back to #1489

adelapena commented:
Thanks for the reviews :)
